### ISSUE TITLE
Any way to leave errors alone and don't kidnap errors in seneca for dev pourposes?

### ISSUE THREAD

**Issue by awill2** [senecajs/seneca#618]
I need that Seneca leave my errors alone while I developing things and let my app die.
Also with seneca.fixedargs.fatal$, Seneca returns a stacktrace, this is bad because i write code in TypeScript and my IDE detect the error in original stack trace of transpiled file and map the line related to .ts file, now Seneca kidnaping my errors and I need to find the error in the transpiled file then find error in the script that I working, it's a hell. There a way to stop Seneca to kidnapping my errors? 
Thanks.

**Comment by rjrodger** [comment]
Hi @awill2  - thanks - could you construct very simple example code to show me what you want to do?

**Comment by awill2** [comment]
just let the script die and throw the fatal error like any js script, don't capture or handle or modify the error in Seneca.
In a dev enviroment this is nice and ok.

**Comment by rjrodger** [comment]
@awill2 all seneca actions are async (they might be remote), so you can't throw errors directly.
I'm sure you have a valid use case :) a few lines of code would really help me to understand it

**Comment by tswaters** [comment]
This may or may not be an incredibly hard thing to solve without breaking changes, or completely tearing apart how seneca does errors currently.

Consider the following example:

```js
process.nextTick(() => {
  foo.baz = 'i screwed up'
})

/*
D:\Code\seneca-tests\test.js:14
  foo.baz = 'i screwed up'
  ^

ReferenceError: foo is not defined
    at process.nextTick (D:\Code\seneca-tests\test.js:14:3)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
    at Module.runMain (module.js:606:11)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

D:\Code\seneca-tests>echo %ERRORLEVEL%
1
*/
```

(FWI, %ERRORLEVEL% is populated with the exit code in windows)

Now do that same thing in an action:

```js
Seneca({log: 'silent'})
  .add('foo:bar', (args, cb) => foo.baz = 'i screwed up')
  .act('foo:bar', (err, result) => {})

/*
D:\Code\seneca-tests>node test.js
(node:4580) DeprecationWarning: 'root' is deprecated, use 'global'

D:\Code\seneca-tests>echo %ERRORLEVEL%
0
*/
```

Or after having called an action --

```js
Seneca({log: 'silent'})
  .add('foo:bar', (args, cb) => cb())
  .act('foo:bar', (err, result) => foo.baz = 'i screwed up')

/*
D:\Code\seneca-tests>node test.js
(node:1016) DeprecationWarning: 'root' is deprecated, use 'global'

D:\Code\seneca-tests>echo %ERRORLEVEL%
0
*/
```

Seneca is most definitely gobbling errors.  For IDEs that can use the results of stderr, stack traces and source maps to point you in the right direction, the above output isn't the most helpful.  

Now let's say I bump up logging a bit to `error` -- 

```
D:\Code\seneca-tests>node test.js
(node:11908) DeprecationWarning: 'root' is deprecated, use 'global'
{"actid":"gq73dgrznrbl/ld62dko2znes","msg":{"foo":"bar","meta$":{"mi":"gq73dgrznrbl","tx":"ld62dko2znes","id":"gq73dgrznrbl/ld62dko2znes","pattern":"foo:bar","action":"(yqu8udvi51wp)","plugin_name":"root$","prior":{"chain":[],"entry":true,"depth":0},"start":1493780676056,"sync":true},"plugin$":{"name":"root$"},"tx$":"ld62dko2znes"},"entry":true,"prior":[],"meta":{"raw":{"foo":"bar"},"plugin_name":"root$","plugin_fullname":"root$","sub":false,"client":false,"args":{"foo":"bar"},"rules":{},"id":"(yqu8udvi51wp)","pattern":"foo:bar","msgcanon":{"foo":"bar"},"priorpath":""},"client":false,"listen":false,"transport":{},"kind":"act","case":"ERR","info":"seneca: Action foo:bar callback threw: foo is not defined.","code":"act_callback","err":{"eraro":true,"orig":{},"code":"act_callback","seneca":true,"package":"seneca","msg":"seneca: Action foo:bar callback threw: foo is not defined.","details":{"message":"foo is not defined","pattern":"foo:bar","instance":"Seneca/79yw67l05fxd/1493780675893/11908/3.3.0/-","orig$":{},"message$":"foo is not defined","plugin":{}},"callpoint":"at Seneca.add.act (D:\\Code\\seneca-tests\\test.js:8:36)"},"duration":32,"level":"error","when":1493780676093}
{"kind":"notice","notice":"hello seneca 79yw67l05fxd/1493780675893/11908/3.3.0/-","level":"info","when":1493780676096}

D:\Code\seneca-tests>echo %ERRORLEVEL%
0
```

That one-line output that github displays doesn't quite do it justice when you're staring at the same thing in a console output and it's quite literally a full wall of JSON output.  And that's the simplest of cases!  Usually there are multiple actions and/or multiple services in play, different services reporting the error, etc.

Now without pulling out a magnifying glass, or copying the results into a json parser can you easily find the error?  I've learned to pick out `callpoint` near the end after much practice, but doing this in a dynamically typed language where errors during development can be quite frequent, and doing it with seneca in the mix makes things difficult and unpleasant.

In short, finding the typo or small problem with the code by looking at the logs that seneca generates is like looking through a needle in a haystack.

**Comment by rjrodger** [comment]
@tswaters this is helpful!

There are a few cases:

- production: you have lots of Seneca instances, and you're pushing stuff up to https://www.elastic.co/webinars/introduction-elk-stack or https://logentries.com/ or whatever - they slurp in the JSON and you can analyse later using their search functionality

- staging: I tend to run a small ELK stack anyway: https://github.com/nodezoo/nodezoo-system/blob/master/stage/nodezoo.yml and just shove the container logs into it

- dev and unit testing: the JSON log output is no good here, so I use `seneca.test('print')` which dumps an abbreviated output, and formats errors:

```
38/c0	act/IN	c6/sp	a:*	{a:1,x:1}
46/c0	act/IN	8q/tx	a:*	{a:2,x:2}
148/s0	act/IN	c6/sp	a:1	{a:1,x:1}
149/s0	act/OUT	c6/sp	a:1	{x:101}
152/s0	act/IN	8q/tx	a:2	{a:2,x:2}
158/s0	act/ERR	8q/tx	a:2	{a:2,x:2}	2	

Error: 2
    at Seneca.<anonymous> (/Users/richard/Projects/seneca/tmp/web.js:13:15)
    at execute_action (/Users/richard/Projects/seneca/seneca.js:1115:20)
    at Object.act_fn [as fn] (/Users/richard/Projects/seneca/seneca.js:1060:11)
    at Immediate.processor (/Users/richard/Projects/seneca/node_modules/gate-executor/gate-executor.js:136:14)
    at runCallback (timers.js:666:20)
    at tryOnImmediate (timers.js:639:5)
    at processImmediate [as _immediateCallback] (timers.js:611:5)

    Action call arguments and location: { a: 2,  x: 2,  'caller$': '\n    Action call arguments and location: { a: 2, x: 2 }\n    at Seneca.<anonymous> (/Users/richard/Projects/seneca/tmp/web.js:31:20)\n',  'timeout$': 22222,  'meta$':    { mi: '8qtprao8ar3n',     tx: 'tx8717xzu03u',     id: '8qtprao8ar3n/tx8717xzu03u',     pattern: 'a:*',     action: '(p7ycfcez01k2)',     plugin_name: 'client$',     plugin_tag: '-',     prior: { chain: [], entry: true, depth: 0 },     start: 1493824757299,     sync: true },  'plugin$': { name: 'root$', tag: undefined },  'tx$': 'dxftl2nr980n' }
    at emitNone (events.js:86:13)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)


92/c0	act/OUT	c6/sp	a:*	{x:101}
```

**Comment by tswaters** [comment]
@rjrodger this is quite helpful. 
Curious, where does `seneca.test('print')` line go - is that an initialization option for the microservice?

**Comment by rjrodger** [comment]
On errors in general, I am updating the error system to pass through Errors rather than wrapping them.

That said, you'll never have the same result as:

```
process.nextTick(() => {
  foo.baz = 'i screwed up'
})
```

That's just a fundamentally different content. Seneca is acting sending and receiving messages, probably from the network. Any individual message error is not a top level error, just something you log. See https://www.joyent.com/node-js/production/design/errors

**Comment by rjrodger** [comment]
@tswaters example code:

```
var Seneca = require('..')

// the server
Seneca({tag:'s0'})
  .test('print')
  .add('a:1', function (msg, reply) {
    reply({x: 100+msg.x})
  })
  .listen()
  .ready(function() {
    var s0 = this

    // the client
    Seneca({tag:'c0'})
      .test('print')
      .client()
      .ready(function () {
        this
          .act('a:1,x:1', this.util.print)
          .ready(function () {
            this.close()
            s0.close()
          })
      })
  })

```
and in unit tests:

```
  it('find', function(done) {
    var seneca = Seneca()
      .test(done) // errors are sent to done, no need to check for them
     //.test(done, 'print') // nice logs
      .add('a:1', function (msg, reply) {
         // decide to respond with an error:
         reply(new Error('foo'))
      })
      .act('a:1', function (ignore, out) {
        // assertions ... done()
      })
  })
```

I'll be making all this a bit easier in 4.x :)
